### PR TITLE
Jmprieur/decoupling token cache serialization and replacing the Cookie cache by a session cache

### PR DESCRIPTION
**Pull request by jmprieur**
## Purpose
Decouples the Token acquisition service based on MSAL.NET and the token cache serialization
This is also a preamble to fixing #19 

## Does this introduce a breaking change?
```
[ ] Yes
[ x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ x] Feature
[ ] Code style update (formatting, local variables)
[x ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2
cd active-directory-aspnetcore-webapp-openidconnect-v2
git checkout jmprieur/decouplingTokenCacheSerialization
Follow the instructions in the README.md to register the application.
build in VS and run
```

### CODE REVIEW

File: Extensions/AuthPropertiesTokenCacheHelper.cs [hunk 1]
```diff
@@ -1,12 +1,57 @@
...
 using Microsoft.Identity.Client;
+using System;
+using System.Security.Claims;
 
 namespace Microsoft.AspNetCore.Authentication
 {
+    /// <summary>
+    /// Extension class enabling adding the CookieBasedTokenCache implentation service
+    /// </summary>
+    public static class CookieBasedTokenCacheExtension
```
Comment (jmprieur) [inline Extensions/AuthPropertiesTokenCacheHelper.cs]: see Startup.cs for the usage of this extension method.

File: Controllers/HomeController.cs [hunk 2]
```diff
@@ -43,7 +43,7 @@ public async Task<IActionResult> Contact()
             var scopes = new string[] { "user.read" };
             try
             {
-                var accessToken = await _tokenAcquisition.GetAccessTokenOnBehalfOfUser(HttpContext, User, scopes);
```
Comment (jmprieur) [inline Controllers/HomeController.cs]: The HttpContext already contains a User (ClaimsPrincipal object), so it's not needed.

File: Extensions/ITokenAcquisition.cs [hunk 3]
```diff
@@ -66,22 +66,13 @@ public interface ITokenAcquisition
...
-
-        /// <summary>
-        /// Gets an access token for a downstream API on behalf of the user of a given account ID
-        /// </summary>
-        /// <param name="accountIdentifier">User account identifier for which to acquire a token. 
-        /// See <see cref="Microsoft.Identity.Client.AccountId.Identifier"/> for a description on how
-        /// to get the account identifier</param>
-        /// <param name="scopes">Scopes for the downstream API to call</param>
-        Task<string> GetAccessTokenOnBehalfOfUser(string userId, string[] scopes);
+        Task<string> GetAccessTokenOnBehalfOfUser(HttpContext context, string[] scopes);
```
Comment (jmprieur) [inline Extensions/ITokenAcquisition.cs]: The ClaimsPrincipal is not needed (User property of HttpContext). 
Also a method just from the userId does not guaranty that the Application is created. Would be nice to expose.

File: Extensions/TokenAcquisition.cs [hunk 4]
```diff
@@ -39,14 +39,17 @@ public class TokenAcquisition : ITokenAcquisition
...
 
+        private ITokenCacheProvider _tokenCacheProvider;
+
         /// <summary>
         /// Constructor of the TokenAcquisition service. This requires the Azure AD Options to 
         /// configure the confidential client application
         /// </summary>
         /// <param name="options">Options to configure the application</param>
-        public TokenAcquisition(IOptionsMonitor<AzureADOptions> options)
+        public TokenAcquisition(IOptionsMonitor<AzureADOptions> options, ITokenCacheProvider tokenCacheProvider)
```
Comment (jmprieur) [inline Extensions/TokenAcquisition.cs]: Dependency injection of the ITokenCacheProvider to decouple the token acquisition from the implementation of the token cache serialization

File: Extensions/TokenAcquisition.cs [hunk 5]
```diff
@@ -132,15 +135,11 @@ public async Task AddAccountToCacheFromAuthorizationCode(AuthorizationCodeReceiv
...
-                var request = context.HttpContext.Request;
-                var currentUri = UriHelper.BuildAbsolute(request.Scheme, request.Host, request.PathBase, request.Path);
-                var credential = new ClientCredential(_azureAdOptions.ClientSecret);
-                Application = new ConfidentialClientApplication(_azureAdOptions.ClientId, currentUri, credential, AuthPropertiesTokenCacheHelper.ForCodeRedemption(context.Properties), null);
-
                 // As AcquireTokenByAuthorizationCodeAsync is asynchronous we want to tell ASP.NET core that we are handing the code
                 // even if it's not done yet, so that it does not concurrently call the Token endpoint.
                 context.HandleCodeRedemption();
+
+                CreateApplicationIfNeeded(context.HttpContext, context.Properties, null);
```
Comment (jmprieur) [inline Extensions/TokenAcquisition.cs]: Refactoring of this common method (code dedupe)

File: Extensions/TokenAcquisition.cs [hunk 6]
```diff
@@ -159,35 +172,53 @@ public async Task AddAccountToCacheFromAuthorizationCode(AuthorizationCodeReceiv
...
-                AuthPropertiesTokenCacheHelper.ForApiCalls(context, AzureADDefaults.CookieScheme), null);
-
-            string userObjectId = user.FindFirst("http://schemas.microsoft.com/identity/claims/objectidentifier").Value;
-            string tenantId = user.FindFirst("http://schemas.microsoft.com/identity/claims/tenantid").Value;
+            CreateApplicationIfNeeded(context, null, AzureADDefaults.CookieScheme);
+            return await GetAccessTokenOnBehalfOfUser(context.User, scopes);
+        }
+
+
+        private static string GetAccountId(ClaimsPrincipal user)
```
Comment (jmprieur) [inline Extensions/TokenAcquisition.cs]: Extracting this method (single responsibility).
Would be actually nice to have in MLSA.NET.

File: appsettings.json [hunk 7]
```diff
@@ -5,6 +5,7 @@
     "TenantId": "[Enter 'common', or 'organizations' or the Tenant Id (Obtained from the Azure portal. Select 'Endpoints' from the 'App registrations' blade and use the GUID in any of the URLs), e.g. da41245a5-11b3-996c-00a8-4d99re19f292]",
     "ClientId": "[Enter the Client Id (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
     "CallbackPath": "/signin-oidc",
+    "SignedOutCallbackPath ": "/signout-callback-oidc",
```
Comment (jmprieur) [inline appsettings.json]: This is to document more easily that the app registration needs to add the sign-out callback (not related to this refactoring)

File: Extensions/ITokenCacheProvider.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        /// <summary>
+        /// Get an MSAL.NET Token cache from the HttpContext, and possibly the AuthenticationProperties and Cookies sign-in scheme
+        /// </summary>
+        /// <param name="httpContext">HttpContext</param>
+        /// <param name="authenticationProperties">Authentication properties (for the cookie based ccookie based token cache serialization). Can be <c>null</c>
+        /// if you don't want to use cookie based token cache serialization</param>
+        /// <param name="signInScheme">[Optional] Authentication properties (for the cookie based cookie based token cache serialization). Can be <c>null</c>
+        /// if you don't want to use cookie based token cache serialization</param>
+        /// <returns>A token cache to use in the application</returns>
+        TokenCache GetCache(HttpContext httpContext, AuthenticationProperties authenticationProperties, string signInScheme = null);
```
Comment (bgavrilMS) [inline Extensions/ITokenCacheProvider.cs]: The argumes of GetCache make this interface very specific. The TailSpin interface is more generic (and also async)

 Task<TokenCache> GetCacheAsync(ClaimsPrincipal claimsPrincipal);

You can still pass in AuthentcaiotnProperties and SignInScheme, but via the constructor.

File: Extensions/ClaimPrincipalExtension.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+namespace Microsoft.AspNetCore.Authentication
+{
+    public static class ClaimsPrincipalExtension
+    {
+        /// <summary>
+        /// Get the Account identifier for an MSAL.NET account from a ClaimsPrincipal
+        /// </summary>
+        /// <param name="claimsPrincipal">Claims principal</param>
+        /// <returns>A string corresponding to an account identifier as defined in <see cref="Microsoft.Identity.Client.AccountId.Identifier"/></returns>
+        public static string GetAccountId(this ClaimsPrincipal claimsPrincipal)
```
Comment (bgavrilMS) [inline Extensions/ClaimPrincipalExtension.cs]: Nice!

File: Extensions/AuthPropertiesTokenCacheHelper.cs [hunk 10]
```diff
@@ -1,12 +1,57 @@
...
+
+        public TokenCache GetCache(HttpContext httpContext, AuthenticationProperties authenticationProperties, string signInScheme)
+        {
+            if (authenticationProperties!=null)
+            {
+                return AuthPropertiesTokenCacheHelper.ForCodeRedemption(authenticationProperties);
+            }
+            else
+            {
+                return AuthPropertiesTokenCacheHelper.ForApiCalls(httpContext, signInScheme?? AzureADDefaults.CookieScheme);
```
Comment (bgavrilMS) [inline Extensions/AuthPropertiesTokenCacheHelper.cs]: I am not sure this is the best place to have this dichotomy. It looks like you are able to get AuthProperties from a SignInScheme, which suggests an adapter pattern. That being said, I do not know if you can invoke the adapt code sooner than on BeforeWrite notification arg.

**Pull request merged**